imap_processing/ena_maps/ena_maps.py

```diff
@@ -867,6 +870,304 @@ def __init__(
+            ..., hp_pix_at_rect_subpix_ctrs
+        ]
+
+        # Weighted mean (weighted by solid angle) of these values over the pixel axis,
+        # which is the last axis of this array
+        weighted_hp_vals_at_rect_pix_ctrs = (
+            hp_vals_at_rect_pix_ctrs * rect_subpix_solid_angle_by_lat
+        )
+        mean_pixel_value = (
+            weighted_hp_vals_at_rect_pix_ctrs.sum(axis=(-1))
+            / full_rect_pixel_solid_angle
+        )
+        return mean_pixel_value
+
+    # Allow for 5 arguments, and self to be passed.
+    # ruff: noqa: PLR0913
+    def get_rect_pixel_value_recursive_subdivs(
+        self,
+        rect_pix_center_lon_lat: np.typing.NDArray | tuple[float, float],
+        rect_pix_spacing_deg: float,
+        value_key: str,
+        tolerances: tuple[float, float] = (1e-3, 1e-12),
+        max_subdivision_depth: int = MAX_SUBDIV_RECURSION_DEPTH,
+    ) -> tuple[np.typing.NDArray, int]:
+        """
+        Recursively subdivide a rectangular pixel to get a mean value within tolerances.
+
+        Takes a rectangular pixel, and recursively breaks it up into
+        smaller and smaller subpixels, then calculates the solid-angle weighted mean
+        of the healpix map's value at this pixel, until the difference
+        between the mean values of two consecutive subdivisions is within the
+        specified tolerances. The function returns the mean value at the final level
+        of subdivision and the depth of recursion.
+
+        Parameters
+        ----------
+        rect_pix_center_lon_lat : np.typing.NDArray | tuple[float, float]
+            The center longitude and latitude of the rectangular pixel.
+        rect_pix_spacing_deg : float
+            The spacing of the rectangular pixel in degrees.
+        value_key : str
+            The name of the value to interpolate from the healpix map.
+        tolerances : tuple[float, float], optional
+            The relative and absolute tolerances for convergence,
+            by default (1e-3, 1e-12).
+        max_subdivision_depth : int, optional
+            The maximum depth of recursion for subdivision,
+            by default MAX_SUBDIV_RECURSION_DEPTH.
+            Computation grows exponentially with depth, but only where the value
+            has a significant gradient between adjacent healpix pixels.
+            If the value is smooth, the recursion depth will be low.
+
+        Returns
+        -------
+        tuple[list[float], int]
+            The mean value at the final level of subdivision and the depth of recursion.
+        """
+        relative_tolerance, absolute_tolerance = tolerances
+
+        # Recursively subdivide a pixel and calculate its mean value until either the
+        # difference between consecutive levels is within the specified tolerances
+        # or the maximum recursion depth is reached
+        depth = 0
+        mean_pixel_value_at_level = []
```

Comment:
Can you keep track of the `prev_mean_pixel_value`, instead of keeping this list? It looks like you only want the current and previous values.

Reply:
Good to know - I'm reading up on hot loops.